src/WebJobs.Script/Host/IShutdownManager.cs

```diff
@@ -0,0 +1,18 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    public interface IShutdownManager
```

Comment:
My concern with this is that we now have 3 different service interfaces floating around for this: the root IApplicationLifetime, IScriptJobHostEnvironment and this new IShutdownManager. This will lead to confusion - e.g. ConsoleScriptJobHostEnvironment still uses IApplicationLifetime directly, but you've refactored WebScriptJobHostEnvironment to use your new interface.

It would be clean and simple I think if you just defined a new EnvironmentInitializationService as a hosted service peer to WebJobsScriptHostService. It can take a direct dependency on IScriptJobHostEnvironment without the circular dependency and you can do your check there?

Reply:
My thought process towards putting this behind a manager is that any service that wants to shut down the app for any reason can just use this directly without the need to worry about any circular dependencies. In that context, ConsoleScriptJobHostEnvironment can also just use the IShutdownManager instead of using the IApplicationLifetime directly. I think the goal was never to use IApplicationLifetime directly and ensure that StopApplication() is called only once even if multiple services call invoke it, right?

And I don't think EnvironmentInitializationService can take IScriptJobHostEnvironment as a direct dependency since IScriptJobHostEnvironment already takes a dependency on IScriptHostManager (WebJobsScriptHostService) so EnvironmentInitializationService will end up depending on WebJobsScriptHostService which is what I'm trying to avoid...

Reply:
Agree with Mathew's comment. Based on the implementation, this also seems redundant. What is the benefit with this new sevice?

Reply:
Yogesh and I discussed yesterday. Given that we've already went down the road of wrapping IApplicationLifetime (with idempotency and other logic), rather than decide to use IApplicationLifetime directly in this case to solve the circular dependency, he's instead factored the wrapper out. I'm OK with that I think, unless we decide to revisit our wrapping completely and just use IApplicationLifetime directly everywhere.

Reply:
I had a quick chat with him about this and I think we have different/better approaches, let's sync offline to discuss.

Reply:
Made changes as discussed offline. Let me know if I missed anything please